// @janelia-flyem/neuroglancer/dist/module/neuroglancer/datasource/python/base.js
/**
 * @license
 * Copyright 2016 Google Inc.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
export var VolumeChunkEncoding;
(function (VolumeChunkEncoding) {
  VolumeChunkEncoding[VolumeChunkEncoding["JPEG"] = 0] = "JPEG";
  VolumeChunkEncoding[VolumeChunkEncoding["NPZ"] = 1] = "NPZ";
  VolumeChunkEncoding[VolumeChunkEncoding["RAW"] = 2] = "RAW";
})(VolumeChunkEncoding || (VolumeChunkEncoding = {}));
export class PythonSourceParameters {}
export class VolumeChunkSourceParameters extends PythonSourceParameters {}
VolumeChunkSourceParameters.RPC_ID = 'python/VolumeChunkSource';
export class MeshSourceParameters extends PythonSourceParameters {}
MeshSourceParameters.RPC_ID = 'python/MeshSource';
export class SkeletonSourceParameters extends PythonSourceParameters {}
SkeletonSourceParameters.RPC_ID = 'python/SkeletonSource';
